test/libsyncengine/update_detection/file_system_observer/testfsoperationset.cpp

```diff
@@ -0,0 +1,523 @@
+        bool functionEndedBeforeUnlock = functionEnded;
+        fsOperationSet._mutex.unlock();
+        Utility::msleep(50);
+        bool functionEndedAfterUnlock = functionEnded;
+        t1.join();
+
+        CPPUNIT_ASSERT_MESSAGE("removeOp (nodeId + type) did not lock the mutex", !functionEndedBeforeUnlock);
+        CPPUNIT_ASSERT_MESSAGE("removeOp (nodeId + type) seems to be deadlock", functionEndedAfterUnlock);
+    }
+
+    // findOp
+    {
+        FSOperationSet fsOperationSet;
+        bool functionEnded = false;
+        fsOperationSet._mutex.lock();
+
+        std::thread t1([&fsOperationSet, &functionEnded]() {
+            FSOpPtr opPtr;
+            functionEnded = false;
+            FSOpPtr res;
+            fsOperationSet.findOp("nodeId", OperationTypeCreate, res);
+            functionEnded = true;
+        });
+
+        Utility::msleep(50);  // Normal delay is < 1 microsecond
+        bool functionEndedBeforeUnlock = functionEnded;
+        fsOperationSet._mutex.unlock();
+        Utility::msleep(50);
+        bool functionEndedAfterUnlock = functionEnded;
+        t1.join();
+
+        CPPUNIT_ASSERT_MESSAGE("findOp did not lock the mutex", !functionEndedBeforeUnlock);
+        CPPUNIT_ASSERT_MESSAGE("findOp seems to be deadlock", functionEndedAfterUnlock);
+    }
+
+    // operator =
+    {
+        FSOperationSet fsOperationSet1;
+        FSOperationSet fsOperationSet2;
+        bool functionEnded = false;
+        fsOperationSet1._mutex.lock();
+        fsOperationSet2._mutex.lock();
+
+        std::thread t1([&fsOperationSet1, &fsOperationSet2, &functionEnded]() {
+            functionEnded = false;
+            fsOperationSet1 = fsOperationSet2;
+            functionEnded = true;
+        });
+
+        Utility::msleep(50);  // Normal delay is < 1 microsecond
+        bool functionEndedBeforeUnlock = functionEnded;
+        fsOperationSet1._mutex.unlock();
+        Utility::msleep(50);  // Normal delay is < 1 microsecond
+        functionEndedBeforeUnlock = functionEndedBeforeUnlock || functionEnded;
+        fsOperationSet2._mutex.unlock();
+        Utility::msleep(50);
+        bool functionEndedAfterUnlock = functionEnded;
+        t1.join();
+
+        CPPUNIT_ASSERT_MESSAGE("operator = did not lock the mutex", !functionEndedBeforeUnlock);
+        CPPUNIT_ASSERT_MESSAGE("operator = seems to be deadlock", functionEndedAfterUnlock);
+    }
+}
+
+void TestFsOperationSet::testCopyConstructor() {
+    FSOperationSet fsOperationSet;
+    NodeId nodeId = "nodeId";
+    char nodeIdSuffix = 0;
+
+    for (OperationType type : _operationTypes) {
+        for (int i = 0; i < 3; i++) {
+            auto op = std::make_shared<FSOperation>(type, nodeId + nodeIdSuffix);
+            nodeIdSuffix++;
+            fsOperationSet.insertOp(op);
+        }
+    }
+
+    FSOperationSet fsOperationSetCopy(fsOperationSet);
+    CPPUNIT_ASSERT_EQUAL(fsOperationSet._ops.size(), fsOperationSetCopy._ops.size());
```

Comment:
Use only the public interface of `FSOperationSet`. You can extend it, if need be.